<a href="https://colab.research.google.com/github/Yamito548/Act-3.2/blob/main/Actividad_3_3_%E2%80%93_IRIS_con_NaiveBayes_y_Cross_Validation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Cross validation

In [62]:
import seaborn as sns
import pandas as pd
import numpy as np
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from matplotlib.colors import ListedColormap
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.metrics import accuracy_score
#Clasificadores
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB, ComplementNB
import matplotlib.pyplot as plt
penguin = sns.load_dataset("penguins")
penguin

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    Male  
1         3800.0  Female  
2         3250.0  Female  
3            NaN     NaN  
4         3450.0  Female  
..           ...     ...  
339          NaN     NaN  
340       4850.0  Female  
341       5750.0    Male  
342       5200.0  Female  
343       5400.0    Male  

[344 rows x 7 columns]

In [63]:
penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']] = \
    penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']].fillna(penguin.mean())

<ipython-input-63-121cb5fe3b6a>:2: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']].fillna(penguin.mean())


In [64]:
le = LabelEncoder()
penguin['sex'] = le.fit_transform(penguin['sex'])

In [65]:
penguin['sex'] = penguin['sex'].fillna(penguin['sex'].mode().iloc[0])

# MultinomialNB

# Truco cross validation, cortito

In [66]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = MultinomialNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))

Resultados de Cross-Validation:
[0.79710145 0.76811594 0.8115942  0.8115942  0.86764706]
Precisión media: 0.81


# GaussianNB

In [67]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = GaussianNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))

Resultados de Cross-Validation:
[0.97101449 0.95652174 0.95652174 0.98550725 0.95588235]
Precisión media: 0.97


# BernoulliNB

In [68]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = BernoulliNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))

Resultados de Cross-Validation:
[0.44927536 0.44927536 0.43478261 0.43478261 0.44117647]
Precisión media: 0.44


# ComplementNB

In [70]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = ComplementNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))

Resultados de Cross-Validation:
[0.69565217 0.71014493 0.69565217 0.66666667 0.73529412]
Precisión media: 0.70


# Cross validation a mano

In [69]:
#Dividamos los datos de entrenamiento
X_train, X_test, y_train, y_test = \
        train_test_split(X, y, test_size=.4, random_state=42)

# Clasificadores
names = ["GaussianNB", "MultiNomialNB", 'BernouilliNB', 'ComplementNB']
classifiers = [GaussianNB(), MultinomialNB(), BernoulliNB(), ComplementNB()]

for name, clf in zip(names, classifiers):

  # Entrenadr el modelo
  clf.fit(X_train, y_train)
  # PREDIT AND SCORE
  score = clf.score(X_test, y_test) #Returns the mean accuracy on the given test data and labels.
  print ("Modelo: %s = %6.2f" % (name, score))

Modelo: GaussianNB =   0.96
Modelo: MultiNomialNB =   0.82
Modelo: BernouilliNB =   0.46
Modelo: ComplementNB =   0.67


En este caso, se puede ver que al usar dos formas diferentes de usar cross validation obtenemos resultados un poco diferentes. Por ejemplo que en el que es más sencillo me dió en complementNB un score de 0.70 mientras que el que es más a mano me dio un 0.67. Siendo un poco mejor, y eso pasa con todos. No creo que esto signifique el 100% es mejor el 1º que el 2º habría que revisar bien como hace los calculos que usa el de cross_val_score